### PR TITLE
Tokenfactory updates

### LINKED ISSUE #1505: [v9] Token Factory
## Background

Token factory should be fully implemented and tested for v9 release.

## Acceptance Criteria

- Review of token factory code
- Create a [more detailed spec](https://github.com/osmosis-labs/osmosis/blob/main/x/tokenfactory/README.md)
- Test token creation on a testnet
- Add a gov parameter to charge a set fee per new denom created
- (optional) Test Cosmwasm binding API for token factory

### PR DESCRIPTION
**Pull request by ValarDragon**
Closes #1505 

## What is the purpose of the change

This PR is doing several updates for the TokenFactory module.

- [x] Rename `nonce` -> `subdenom`
- [x] Enforce maximum `subdenom` length (and derive it)
- [ ] Fix API in the bindings due to ChangeAdmin functionality
- [ ] Add more methods to the TokenFactroy bindings API

## Brief Changelog

* Rename `nonce` -> `subdenom`
* Enforce maximum `subdenom` length (and derive it)
* Fix API in the bindings due to ChangeAdmin functionality
* Add more methods to the TokenFactroy bindings API
## Testing and Verifying
(TODO)

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? TODO
  - How is the feature or change documented? Specification and struct godocs

### CODE REVIEW

**Comment by ValarDragon** [comment]
Go binding structs done, still need to update all tests

File: app/wasm/test/custom_msg_test.go [hunk 1]
```diff
@@ -70,13 +70,14 @@ func TestMintMsg(t *testing.T) {
 	}}
 	err := executeCustom(t, ctx, osmosis, reflect, lucky, msg, sdk.Coin{})
 	require.NoError(t, err)
+	sunDenom := fmt.Sprintf("factory/%s/%s", lucky.String(), msg.CreateDenom.Subdenom)
```
Comment (ValarDragon) [inline app/wasm/test/custom_msg_test.go:73]: This is code smell that this isn't trivial to get.

Need to see how it looks inside the cosmwasm.

**Comment by ValarDragon** [comment]
Remaining TODO's:
- [ ] Add receiver code for BurnTokens (Be sure to handle complexity around Burn From Address being "")
- [ ] Add receiver code for ChangeAdmin
- [ ] Add tests for the above two
- [ ] Update cosmwasm contracts
- [ ] Ensure its easy to get the generated denom from within cosmwasm. (Eliminate need for doing sprintf's yourself)

File: app/wasm/bindings/msg.go [hunk 2]
```diff
@@ -4,28 +4,47 @@ import sdk "github.com/cosmos/cosmos-sdk/types"
...
+	Amount        sdk.Int `json:"amount"`
+	MintToAddress string  `json:"mint_to_address"`
+}
+
+type BurnTokens struct {
+	Denom  string  `json:"denom"`
+	Amount sdk.Int `json:"amount"`
+	// BurnFromAddress must be set to "" for now.
+	BurnFromAddress string `json:"burn_from_address"`
 }
```
Comment (ValarDragon) [inline app/wasm/bindings/msg.go:48]: Need to decide if Rust side we include BurnFromAddress and require it being empty on golang side or keep it empty for now.

The go side can handle either way. Not sure if bindings in Rust use traits vs macros vs direct structs. For traits / macros we can remove it and add it in optionally later. For a direct struct, depends on if fields are private and its generated via constructor, or if fields are public and can be directly put in.

Comment (sunnya97) [reply app/wasm/bindings/msg.go:48]: I think we should add it, and then throw on the Go side

**Comment by ValarDragon** [comment]
Existing tests and bindings fixed

Need to work on receivers for the new messages

File: app/wasm/bindings/query.go [hunk 3]
```diff
@@ -22,8 +22,8 @@ type OsmosisQuery struct {
 }
 
 type FullDenom struct {
-	Contract string `json:"contract"`
-	SubDenom string `json:"sub_denom"`
+	CreatorAddr string `json:"creator_addr"`
+	Subdenom    string `json:"subdenom"`
```
Comment (alexanderbez) [inline app/wasm/bindings/query.go:26]: Why is this `Subdenom` and not `SubDenom`?

Comment (ValarDragon) [reply app/wasm/bindings/query.go:26]: subdenom seems like a better name right? We say subroutine not subRoutine, or sub_routine

File: app/wasm/bindings/msg.go [hunk 4]
```diff
@@ -4,28 +4,49 @@ import sdk "github.com/cosmos/cosmos-sdk/types"
...
 	/// Swap over one or more pools
 	Swap *SwapMsg `json:"swap,omitempty"`
 }
 
+/// CreateDenom creates a new factory denom, of denomination:
+/// factory/{creating contract address}/{Subdenom}
+/// Subdenom can be of length at most 44 characters, in [0-9a-zA-Z./]
+/// The (creating contract address, subdenom) pair must be unique.
+/// The created denom's admin is the creating contract address,
+/// but this admin can be changed using the UpdateAdmin binding.
```
Comment (ValarDragon) [inline app/wasm/bindings/msg.go:27]:
```suggestion
/// but this admin can be changed using the ChangeAdmin binding.
```

Comment (ValarDragon) [reply app/wasm/bindings/msg.go:27]: Its called change admin in go

**Comment by p0mvn** [comment]
receiver code for ChangeAdmin: https://github.com/osmosis-labs/osmosis/pull/1600

**Comment by ValarDragon** [comment]
Thoughts on merging this to main now, given other work happening concurrently?

**Pull request merged**